Source/EasyNetQ/Consumer/ConsumerDispatcher.cs

```diff
@@ -8,53 +8,68 @@ namespace EasyNetQ.Consumer
         private readonly ILog logger = LogProvider.For<ConsumerDispatcher>();
-        private readonly BlockingCollection<Action> queue;
-        private bool disposed;
+        private readonly CancellationTokenSource cancellation = new CancellationTokenSource();
+        private readonly SemaphoreSlim actionsAvailable = new SemaphoreSlim(0);
+        private readonly ConcurrentQueue<Action> durableActions = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Action> transientActions = new ConcurrentQueue<Action>();
 
         public ConsumerDispatcher(ConnectionConfiguration configuration)
         {
             Preconditions.CheckNotNull(configuration, "configuration");
 
-            queue = new BlockingCollection<Action>();
-
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
-                {
+                while (!cancellation.IsCancellationRequested || !durableActions.IsEmpty || !transientActions.IsEmpty)
                     try
                     {
-                        action();
+                        if (durableActions.TryDequeue(out var action) || transientActions.TryDequeue(out action))
+                            action();
+                        else
+                            actionsAvailable.Wait(cancellation.Token);
+                    }
+                    catch (OperationCanceledException)
+                    {
                     }
                     catch (Exception exception)
                     {
                         logger.ErrorException(string.Empty, exception);
                     }
-                }
             }) {Name = "EasyNetQ consumer dispatch thread", IsBackground = configuration.UseBackgroundThreads};
             thread.Start();
         }
 
         public void QueueAction(Action action)
+        {
+            QueueAction(action, false);
+        }
+
+        public void QueueAction(Action action, bool surviveDisconnect)
         {
             Preconditions.CheckNotNull(action, "action");
-            queue.Add(action);
+
+            if(cancellation.IsCancellationRequested)
+                throw new InvalidOperationException("Consumer dispatcher is stopping or already stopped");
+
+            if (surviveDisconnect)
+                durableActions.Enqueue(action);
+            else
+                transientActions.Enqueue(action);
+
+            actionsAvailable.Release();
         }
 
         public void OnDisconnected()
         {
             // throw away any queued actions. RabbitMQ will redeliver any in-flight
             // messages that have not been acked when the connection is lost.
-            while (queue.TryTake(out _))
+            while (transientActions.TryDequeue(out _))
             {
             }
         }
 
         public void Dispose()
         {
-            queue.CompleteAdding();
-            disposed = true;
+            cancellation.Cancel();
```

Comment:
Should we wait for a join of thee dispatcher thread here?

Reply:
We can wait to eliminate possible side effects but I am not sure. By the way excellent decision with cancellation token. I am ashamed that I myself did not think about it.

Reply:
Also if you prefer to wait you can dispose `actionsAvailable`.